### DISCUSSION TITLE
Controller render field is undefined

### DISCUSSION THREAD

**Discussion by chiararv** [react-hook-form/react-hook-form#4835] [Q&A]
I'm getting an error when using onChange, value and onBlur because field (inside render prop) seems to be undefined.
I recently migrated from v6 to v7. Until today it worked just fine.
This is my code:

 <Controller
      control={control}
      defaultValue={defaultValue}
      name={name}
      rules={{
        required: { value: required, message: ERROR_MESSAGES.REQUIRED },
        pattern: regExPattern,
        minLength: minLengthValidation,
      }}
      render={({ field: { onChange, onBlur, value } }) => (
        <Input
          label={label}
          value={value}
          onBlur={onBlur}
          error={errors}
          helperText={helperText}
          onChange={text => {
            onChange(text);
          }}
          isNumber={isNumber}
          rightIcon={rightIcon}
          maxLength={maxLength}
          {...restProps}
        />
      )}
    />
![Screenshot_20210416-121103](https://user-images.githubusercontent.com/62294095/115045779-345d6980-9ead-11eb-98e4-be8395b8fdbd.png)

**Comment by mbretter** [comment]
which version (6, 7) are you using?

**Comment by chiararv** [comment]
I recently migrated from v6 to v7. Until today it worked just fine. If I use it as I did before in v6 it works.
It's just field that doesn't work

**Comment by bluebill1049** [comment]
can you log the entire `field` object?

**Reply by chiararv** [reply]
Sorry I didn't reply sooner, had some issues related with corona virus. The problem went away just as misteriously as it came. Field is working fine now.

**Comment by mengqing** [comment]
I'm using 7.3.6 and is also getting the same issue. It seems that on SSR the render prop returns correct objects (`field`, `fieldState`, `formState`), however on CSR, the return type for the render prop is the `field` object

**Reply by bluebill1049** [reply]
Can u reproducr that in a codesandbox?

**Comment by sellsjonathan** [comment]
I have an component that composes the `<Controller>` component with labels and validations messages etc., and I pass an MUI input to its `render` prop like this:
```jsx
<Controller
        name={name}
        control={useFormControl}
        defaultValue={defaultValue}
        render={({ field }) => {
          console.log('FIELD TEST', field);
          return <InputComponent {...field} {...rest} />;
        }}
      />
  ```
Notice, this component also has a `defaultValue` prop, which will default to `undefined`

When appending a field through `useFieldArray` method like so:

` const handleAddRule = (index: number) => append({ name: 'New rule' });`

console logs => `FIELD TEST {name: "rules.1.name", value: undefined, onChange: ƒ, onBlur: ƒ, ref: ƒ}`

The new field gets added, at the correct location within my form schema, but its value gets immediately overwritten, by what I believe is the defaultValue passed to the `<Controller>` component.  

Am I missing something? or should I be setting the defaultValue somehow?

I noticed a similar issue potentially got fixed in (#4861), but then reverted in (#4945).

Thanks for your time.

**Reply by mbretter** [reply]
remove the defaultValue property from the Controller and set it using useForm({defaultValues:{.....})
I have had the same issue, this seems to be a problem if both defaultValues are present, at useForm and at the Controller.

**Reply by bluebill1049** [reply]
that's right, it's either. 

<img width="963" alt="Screen Shot 2021-05-02 at 5 06 55 pm" src="https://user-images.githubusercontent.com/10513364/116805307-cda4a500-ab68-11eb-9c3f-3ab677ef41aa.png">

**Reply by sellsjonathan** [reply]
Hey, thanks for your replies.  

If I don't set `defaultValue` manually at the field level through the `defaultValue` prop, then this prop will still be `undefined` for the `<Controller>`.  

**So I believe it's still a bug** when using `useFieldArray` methods (`append`, `insert` etc.) and trying to set a default value for the dynamically added fields with a `<Controller>` component.

**EG:**
```jsx
<Controller
        name={name}
        control={useFormControl}
        // defaultValue={defaultValue}
        render={({ field }) => {
          console.log('FIELD TEST', field);
          return <InputComponent {...field} {...rest} />;
        }}
      />
   ```

Still console.logs => `FIELD TEST {name: "rules.1.name", value: undefined, onChange: ƒ, onBlur: ƒ, ref: ƒ}`
      

I can confirm that defaultValues get set properly through the `useForm` param - but that doesn't help in this usecase when you're adding fields to the schema via `useFieldArray` methods.

Thoughts?

**Reply by mbretter** [reply]
You are right, I'm adding a default object, my defaults looks like this:
```JS
{
        name: "",
        order: 10,
        type: "new",
        filters: [{
          id: null,
          target: null,
          field: null,
          op: null,
          value: null
        }],
        active: true,
    }
```
this seems to be enough, for getting useFieldArray to work.

```JS
    const {fields: filters, append: appendFilter, remove: removeField} = useFieldArray({
        control: rhf.methods.control,
        name: "filters",
        keyName: "id"
    });
```
IMHO, the clean solution would be the possibility for supplying defaultValues to useFieldArray
```JS
    const {fields: filters, append: appendFilter, remove: removeField, defaultValues: {id: null,
          target: null,
          field: null,
          op: null,
          value: null}} = useFieldArray({
        control: rhf.methods.control,
        name: "filters",
        keyName: "id"
    });
```

**Reply by sellsjonathan** [reply]
@bluebill1049  Do you agree that this is a bug to flag? Let me know if you need more detail.

**Reply by bluebill1049** [reply]
It's a design limitation with field array, as we need default value to overwrite controller' default value during shuffle and reorder (for the first render)

**Reply by bluebill1049** [reply]
Also can u supply a codesandbox with ur problem.

**Reply by sellsjonathan** [reply]
Sure thing,

https://codesandbox.io/s/react-hook-form-usefieldarray-bug-hb4ny

Comments in file:
As you can see, the `defaultValue` does not get added to the form schema on `append()`, or any other method of `useFieldArray`

If you try to type in the dynamic fields after they are rendered you'll see the react warning:

```Warning: A component is changing an uncontrolled input to be controlled. This is likely caused by the value changing from undefined to a defined value, which should not happen. Decide between using a controlled or uncontrolled input element for the lifetime of the component. ```

This is because the value of these dynamic fields are defaulting to that of the  `<Controller>`'s `defaultValue` which is `undefined` (when unset), and upon user input, the value gets updated, to a string value.

 **BUG**
`useFieldArray` methods, are either not setting the value of the newly created field in the fieldArray, or there is some sort of race condition, on rendering inputs via `<Controller>` that quickly sets the value of the newly created input to the `<Controller>`'s `defaultValue`

**Reply by bluebill1049** [reply]
Where is default value at controller for useFieldArray?

**Reply by sellsjonathan** [reply]
It's not set on the `<Controller>` so it will be `undefined`.

My assumption is that the the parameter in `append()`, should set the value of the input that is being added to the field array.

**Reply by bluebill1049** [reply]
We will need defaultValue for useFieldArray.

**Reply by sellsjonathan** [reply]
If you use defaultValue in this way, you'll be unable to use reset() to set defaultValues for the fieldArray once data comes back from a data source, for an edit form flow - since as you previously mentioned, you use one method or the other.

Surely the parameter for the useFieldArray methods are supposed to set the the defaultValue...?  No? 

```jsx
    <Button
       variant="contained"
       onClick={() => append("Should show this default value on append")}
     >
   ```

**Reply by bluebill1049** [reply]
I don't follow...

**Reply by mbretter** [reply]
I have modified the CSB to get it working:
https://codesandbox.io/s/react-hook-form-usefieldarray-bug-forked-nknkm?file=/src/App.js

This shows the problem with useFieldArray and the defaultValues:
* in real life we are using some kind of async requests for fetching the data, therefore we are using reset(), we must provide the defaultValues at useForm level, not at the Controller
* useFieldArray needs defaultValues to be able to set the some kind of values on the newly created fields, but it has no information where to get them, because the sub-array (test in this case) is empty or does not exists.
* In order to get this working, we have to set a default object/value into the sub-array, but this causes, that we always have at least one entry on the form.
* If we are getting from the server back an empty list (defaultValues: { name: "Joe Biden", test: [] }), there are no defaultValues for useFieldArray available then.

**Reply by sellsjonathan** [reply]
@mbretter  , Ok this is covers all cases by the looks of it, good job.

Do you see this as a workaround, or the actual pattern to follow?

**Reply by mbretter** [reply]
It is simply, when using useFieldArray I would not provide the defaultValues at useForm, I would use the defaultValue at Controller.
The important thing is to retrieve the defaultValue from the field:
`defaultValue={field.whatever || ""}`
initially, this was my fault.

I would not see this as workaround, for me this is the way to go ... ;)

**Reply by sellsjonathan** [reply]
Well good thing I wrote up all this text and took up your time for a simple `||` haha.

Seriously though, thanks to both of you for being so responsive and helpful.

**Reply by bluebill1049** [reply]
you are welcome @sellsjonathan and thank you very much @mbretter for stick around and help.

**Comment by bluebill1049** [comment]
Thanks @mbretter 

> in real life we are using some kind of async requests for fetching the data, therefore we are using reset(), we must provide the defaultValues at useForm level, not at the Controller

I am not sure why we must provide `defautValues`.

> If we are getting from the server back an empty list (defaultValues: { name: "Joe Biden", test: [] }), there are no defaultValues for useFieldArray available then.

If it's an empty array, then it would not create a new `useFieldArray`, can you get around the issue with custom `register` at `useEffect`?

**Reply by mbretter** [reply]
> I am not sure why we must provide `defautValues`.

yes, I am not sure too ;)

I have modified the CSB
https://codesandbox.io/s/react-hook-form-usefieldarray-bug-forked-nknkm?file=/src/App.js
it now does not use any defaults at useForm, generic defaults are provided strictly at the Controller.

The example is now working perfectly, but this means, that defaultValues at useForm and useFieldArray will not really work well.

**Reply by bluebill1049** [reply]
<img width="493" alt="Screen Shot 2021-05-05 at 5 15 43 pm" src="https://user-images.githubusercontent.com/10513364/117107802-8830e380-adc5-11eb-864d-bc325dac3bcf.png">

shouldn't this be assigned with `field.xxx`?

**Reply by mbretter** [reply]
> shouldn't this be assigned with `field.xxx`?

if we are dealing with async requests, we have the form displayed, but we do not have any values available yet, we need a generic default. After the data has been loaded we are using reset() to set the data, or am I missing something?

**Reply by bluebill1049** [reply]
another issue: name={`test.${index}`} we don't support flat array as well.

**Reply by mbretter** [reply]
I have modified the CSB:
https://codesandbox.io/s/react-hook-form-usefieldarray-bug-forked-nknkm?file=/src/App.js
It has a nested object now.

I know now why I have removed the defaultValue from the Controller, because reset() is not working anymore then.
Click on the Reset/Simulate button, AAAAA should be displayed, but it is not.
When clicking twice on the reset button AAAAA is shown?!

**Reply by bluebill1049** [reply]
`defaultValue={field.whatever}` ?

**Reply by mbretter** [reply]
ok next round, modified again:
https://codesandbox.io/s/react-hook-form-usefieldarray-bug-forked-nknkm?file=/src/App.js
we have an empty form at the beginning, click an add dynamic field, enter a value:
Warning: A component is changing an uncontrolled input to be controlled. 

`defaultValue={field.whatever || ""}`
does the job.

The rest of the stuff seems to be working now.

So finally, useFieldArray and defaultValue at the Controller level is working.
